packages/metro-memory-fs/src/index.js

```diff
@@ -675,19 +675,22 @@ class MemoryFs {
     }
   };
 
-  mkdirSync: (dirPath: string | Buffer, mode?: number) => void = (
+  mkdirSync: (
     dirPath: string | Buffer,
-    mode?: number,
+    mode?: number | {recursive?: boolean, mode?: number},
+  ) => void = (
+    dirPath: string | Buffer,
+    mode?: number | {recursive?: boolean, mode?: number},
   ): void => {
-    if (mode == null) {
-      mode = 0o777;
-    }
+    const recursive = typeof mode != 'number' && mode?.recursive;
+    const realMode = (typeof mode == 'number' ? mode : mode?.mode) ?? 0o777;
+
     dirPath = pathStr(dirPath);
     const {dirNode, node, basename} = this._resolve(dirPath);
-    if (node != null) {
+    if (node != null && !recursive) {
       throw makeError('EEXIST', dirPath, 'directory or file already exists');
     }
-    dirNode.entries.set(basename, this._makeDir(mode));
+    dirNode.entries.set(basename, this._makeDir(realMode));
```

Comment:
can we move the initialization of `realMode` here? we potentially may not even use the realMode we calculated if the condition in L690 passes.

Reply:
I could. I think keeping logic like this based on the input of a function together at the start helps the readability of the code. Just as the original code had the default value of `mode` provided at the top of the file and not down near `this._makeDir(...)` and just as the original code did not make use of the updated `mode` value unless the if check were violated.

Given that this would optimize at best a degenerate case of the node already existing and at worst 2 more comparisons compared to the prior version, I don't know that its worth the legibility change. But if you insist I will make that change.

(Side note, I don't know flow. In TS, I could have reused the `mode` variable in my default processing assignment instead of introducing `realMode` and the compiler would infer that from that point on the code is always a number and not a number or undefined or object. Is this possible with flow? See here for TS example: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAWwNYBMYCcDKBPSACgChEzFMsAFAQygAsAuRAZyixjAHMAaU85HHQBTAPzMwIZACNhWRAB9EAbyzCIILCxgA3Mc2lw4AG2E0wPFEP2JJMuQF8+ZAJQr+ZCAjaI1NYwCy1ogAvIiEUHgADsJwwFYioWEA5HayWMmIognCiMyCIqIAdAXCbqLZAAxwAOx1ANweOaHhkTFxzSEpaXKZ2aV5OcWl5VW1DcRNXmAsJsJFxnBchCON5IgA9BvrO4gAeqL8DkA )